MdeModulePkg/Library/UefiBootManagerLib/BmBoot.c

```diff
@@ -2436,6 +2436,241 @@ BdsLibGetBootableHandle (
+          FullOptionName, ConvertDevicePathToText(OptDevicePath, FALSE, FALSE)
+          ));
+
+  Status = EfiBootManagerInitializeLoadOption (
+             &BootOptions[(*BootOptionCount)++],
+             LoadOptionNumberUnassigned,
+             LoadOptionTypeBoot,
+             LOAD_OPTION_ACTIVE,
+             FullOptionName,
+             OptDevicePath,
+             NULL,
+             0
+             );
+  ASSERT_EFI_ERROR (Status);
+  FreePool (FullOptionName);
+  FreePool (OptDevicePath);
+
+  if (Description != NULL)
+    FreePool (Description);
+
+  return BootOptions;
+}
+
+EFI_BOOT_MANAGER_LOAD_OPTION *
+BmEnumeratePreInstalledBootOptions (
+  IN OUT UINTN                                 *BootOptionCount
+  )
+{
+  EFI_STATUS                            Status;
+  UINTN                                 HandleCount;
+  EFI_HANDLE                            *Handles;
+  EFI_BLOCK_IO_PROTOCOL                 *BlkIo;
+  UINTN                                 Index;
+  UINTN                                 OsIdx;
+  UINTN                                 OptionsArraySize;
+  EFI_IMAGE_DOS_HEADER                  DosHeader;
+  EFI_IMAGE_OPTIONAL_HEADER_UNION       HdrData;
+  EFI_IMAGE_OPTIONAL_HEADER_PTR_UNION   Hdr;
+  PRE_INSTALLED_BOOT_OPT                DtsBootOpt;
+  EFI_BOOT_MANAGER_LOAD_OPTION          *BootOptions;
+
+  ASSERT (BootOptionCount != NULL);
+
+  BootOptions = NULL;
+
+  //
+  // Parse simple file system not based on block io
+  //
+  gBS->LocateHandleBuffer (
+         ByProtocol,
+         &gEfiSimpleFileSystemProtocolGuid,
+         NULL,
+         &HandleCount,
+         &Handles
+         );
+  for (Index = 0; Index < HandleCount; Index++) {
+
+    DEBUG ((EFI_D_INFO, "%a: Processing file system:\n  %s\n", __FUNCTION__,
+          ConvertDevicePathToText(DevicePathFromHandle (Handles[Index]), FALSE, FALSE)
+          ));
+
+    //
+    // Skip the removable media, except if DTS
+    //
+    Status = gBS->HandleProtocol (
+                    Handles[Index],
+                    &gEfiBlockIoProtocolGuid,
+                    (VOID **) &BlkIo
+                    );
+    if (!EFI_ERROR (Status) && BlkIo->Media->RemovableMedia) {
+      Hdr.Union = &HdrData;
+      Status = BdsLibGetImageHeader (
+                 Handles[Index],
+                 DtsBootOpt.FileName,
+                 &DosHeader,
+                 Hdr
+                 );
+      if (!EFI_ERROR (Status) &&
+        EFI_IMAGE_MACHINE_TYPE_SUPPORTED (Hdr.Pe32->FileHeader.Machine) &&
+        Hdr.Pe32->OptionalHeader.Subsystem == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION) {
+        BootOptions = CreatePreInstalledBootOption(
+                        BootOptions,
+                        BootOptionCount,
+                        Handles[Index],
+                        &DtsBootOpt
+                        );
+      } else {
+        continue;
+      }
+    }
+
+    OptionsArraySize = sizeof (PreInstalledBootOpts) / sizeof(PRE_INSTALLED_BOOT_OPT);

```

Comment:
There is `ARRAY_SIZE()` macro which should work here and for the next loop below.

Reply:
Using ARRAY_SIZE https://github.com/Dasharo/edk2/pull/70/commits/d9d411152e4b6244f44d4f5af663b63145b6e4f6#diff-d2820fe49b39a17c1aa46361e3d22ee285be3bd600dae98c0dc36d3bf030d04fR2746